moveit_core/online_signal_smoothing/src/acceleration_filter.cpp

```diff
@@ -0,0 +1,145 @@
+
+/* Author: Paul Gesel
+   Description: Applies smoothing by limiting the  acceleration between consecutive commands
+ */
+
+#include <moveit/online_signal_smoothing/acceleration_filter.h>
+#include <rclcpp/clock.hpp>
+#include <rclcpp/logging.hpp>
+
+// Disable -Wold-style-cast because all _THROTTLE macros trigger this
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+
+namespace online_signal_smoothing
+{
+
+bool AccelerationLimitedPlugin::initialize(rclcpp::Node::SharedPtr node, moveit::core::RobotModelConstPtr /* unused */,
+                                           size_t num_joints)
+{
+  node_ = node;
+  num_joints_ = num_joints;
+
+  cur_velocity_ = Eigen::VectorXd::Zero(num_joints);
+  prev_velocity_ = Eigen::VectorXd::Zero(num_joints);
+  cur_acceleration = Eigen::VectorXd::Zero(num_joints);
+
+  auto param_listener_ = online_signal_smoothing::ParamListener(node_);
+  auto params = param_listener_.get_params();
+  joint_acceleration_limit_ = params.joint_acceleration_limit;
+  update_timeout_ = params.update_timeout;
+
+  return true;
+}
+
+bool AccelerationLimitedPlugin::doSmoothing(Eigen::VectorXd& positions, Eigen::VectorXd& /* unused */,
+                                            Eigen::VectorXd& /* unused */)
+{
+  const long num_positions = positions.size();
+  if (num_positions != num_joints_)
+  {
+    RCLCPP_ERROR_THROTTLE(
+        node_->get_logger(), *node_->get_clock(), 1000,
+        "The length of the joint positions parameter is not equal to the number of joints, expected %zu got %zu.",
+        num_joints_, num_positions);
+    return false;
+  }
+  else if (last_positions_[0].first.size() != num_joints_)
+  {
+    RCLCPP_ERROR_THROTTLE(node_->get_logger(), *node_->get_clock(), 1000,
+                          "Reset must be called before doSmoothing for the AccelerationLimitedPlugin plugin.");
+    return false;
+  }
+
+  if (!last_positions_[0].second.has_value())
+  {
+    positions = last_positions_[0].first;
+    last_positions_[0].second = node_->now();
+  }
+  else if (!last_positions_[1].second.has_value())
+  {
+    positions = last_positions_[1].first;
+    last_positions_[1].second = node_->now();
+  }
+  else
+  {
+    rclcpp::Time cur_time = node_->now();
+    double dt_1 = (last_positions_[1].second.value() - last_positions_[0].second.value()).seconds();
+    double dt_2 = (cur_time - last_positions_[1].second.value()).seconds();
+    if (dt_1 > update_timeout_ || dt_2 > update_timeout_)
+    {
+      RCLCPP_ERROR_THROTTLE(node_->get_logger(), *node_->get_clock(), 1000,
+                            "The maximum allowed time between filter updates exceeded! The `doSmoothing` method must "
+                            "be called at a higher rate.");
+      return false;
+    }
+
+    prev_velocity_ = (last_positions_[1].first - last_positions_[0].first) / dt_1;
+    cur_velocity_ = (positions - last_positions_[1].first) / dt_2;
+    cur_acceleration = (cur_velocity_ - prev_velocity_) / dt_2;
```

Comment:
We're not checking that `dt_1` and `dt_2` are greater than zero, so we run a risk of division by zero here